src/SeoToolkit.Umbraco.Redirects.Core/Helpers/RedirectsImportHelper.cs

```diff
@@ -73,132 +73,140 @@ private bool UrlExists(string oldUrl)
+                    var urlValidation = ValidateRedirectUrl(fromUrl, parsedData, parser.LineNumber);
+                    if (!urlValidation.Success)
                     {
-                        return Attempt<Dictionary<string,string>?, string>.Fail($"Redirect already exists for 'from' URL: {fromUrl} validation aborted.", result: null);
+                        return urlValidation;
                     }
-                    if (parsedData.ContainsKey(fromUrl.TrimEnd("/")))
+
+                    var codeValidation = ValidateRedirectCode(redirectCode, fromUrl);
+                    if (!codeValidation.Success)
                     {
-                        return Attempt<Dictionary<string,string>?, string>.Fail($"Url appears more then one time in import file: {fromUrl}", result: null);
+                        return codeValidation;
                     }
-                    parsedData.Add(fromUrl, toUrl);
 
+
+                    parsedData.Add(CreateRedirect(fromUrl, toUrl, redirectCode));
                 }
                 else
                 {
-                    return Attempt<Dictionary<string,string>?, string>.Fail($"line {parser.LineNumber}", result: null);
+                    return Attempt<List<Redirect>, string>.Fail($"line {parser.LineNumber}", result: null);
                 }
             }
-            return Attempt<Dictionary<string,string>?, string>.Succeed(string.Empty, parsedData);
-        }
-
 
+            return Attempt<List<Redirect>, string>.Succeed(string.Empty, parsedData);
+        }
     }
-     private Attempt<Dictionary<string,string>?, string> ValidateExcel(Stream fileStream)
+
+    private Attempt<List<Redirect>, string> ValidateExcel(Stream fileStream)
     {
         fileStream.Position = 0;
         Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
-
         try
         {
             using var reader = ExcelReaderFactory.CreateReader(fileStream);
             var result = reader.AsDataSet();
             var dataTable = result.Tables[0];
-
-            var parsedData = new Dictionary<string, string>();
-
+            var parsedData = new List<Redirect>();
             for (var i = 0; i < dataTable.Rows.Count; i++)
             {
                 var row = dataTable.Rows[i];
-                if (row.ItemArray.Length != 2)
+                if (row.ItemArray.Length is not (2 or 3))
                 {
-                    return Attempt<Dictionary<string, string>?, string>.Fail($"only 2 columns allowed on row {i + 1}");
+                    return Attempt<List<Redirect>, string>.Fail($"2 or 3 columns required on row {i + 1}");
                 }
 
                 var fromUrl = CleanFromUrl(row[0].ToString());
                 var toUrl = Uri.IsWellFormedUriString(row[1].ToString(), UriKind.Absolute)
                     ? row[1].ToString()
                     : row[1].ToString()?.EnsureEndsWith("/").ToLower();
-
+                var redirectCode = GetRedirectCode(row[2].ToString());
```

Comment:
I think this requires another check on the length right? You allow it to still be 2 on line 160, but if the length is 2 it will fail on this line

Reply:
Example with a row that has only 2 fields filled in:
![image](https://github.com/user-attachments/assets/1b4df48b-5386-41af-80a0-24c54150b8d1)

row[2].toString() nicely returns an string.Empty even though row[2] is null.
![image](https://github.com/user-attachments/assets/652f0e8a-6633-434b-a14c-ae4f2e3f88fe)